### PR TITLE
Proposed change for Focus Navigation

### PR DESCRIPTION
**Pull request by TakayoshiKochi**
As discussed in
https://github.com/w3c/webcomponents/issues/375

This adds a clarification about focus navigation order for
nodes distributed under a slot element.

Also added handling tabindex="-1" case.

### CODE REVIEW

**Comment by TakayoshiKochi** [comment]
@hayatoito PTAL

File: spec/shadow/index.html [hunk 1]
```diff
@@ -1139,6 +1141,27 @@
           </li>
         </ol>
 
+        <p>Likewise, the <a>sequential focus navigation order</a> for <a>distributed nodes</a>
```
Comment (hayatoito) [inline spec/shadow/index.html]: Wrong... This should not be _distributed nodes_. This should be _assigned nodes_.

In addition to that, you should consider the followings:
1. Do not forget to include descendants of assigned nodes into this focus navigation group.
2. Consider the fallback contents of slots, in addition to assigned nodes.
3. When you try to consider 1 and 2, do not forget the case where a slot is a descendant of another slot.
4. The traversal order is not defined for _sequential focus navigation order for distributed nodes_. Define the traversal order of this focus navigation group.
5. Exclude an assigned node (and more) from its root node's focus navigation group. One node should not join more than one focus navigation group.
6. Mention how tabindex values are _scoped_ at each focus navigation order.

File: spec/shadow/index.html [hunk 2]
```diff
@@ -1139,6 +1141,27 @@
...
+            <strong>must</strong> be inserted in place of <var>A</var> as if <var>A</var>
+            were assigned the <a>tabindex</a> value 0 for determining its position.</li>
+          <li>Otherwise the <a>sequential focus navigation order</a> for its
+            <a>distributed nodes</a> <strong>must</strong> be inserted in place of
+            <var>A</var> according to its <a>tabindex</a> value.</li>
+        </ol>
+        <p class="note">When a <a>slot</a> <var>A</var> is assigned to another
+          <a>slot</a> <var>B</var>,
+          the <a>sequential focus navigation order</a> of <var>A</var>'s <a>distributed nodes</a>
+          is inserted into <var>B</var>'s <a>sequential focus navigation order</a> following
```
Comment (hayatoito) [inline spec/shadow/index.html]: We do not need to mention this.
Please try to have a general rule which covers this case.

File: spec/shadow/index.html [hunk 3]
```diff
@@ -1139,6 +1141,27 @@
...
+            <strong>must</strong> be skipped if A is assigned <a>tabindex</a> value -1.
+          <li>If <var>A</var> is not given <a>tabindex</a>,
+            The <a>sequential focus navigation order</a> for its <a>distributed nodes</a>
+            <strong>must</strong> be inserted in place of <var>A</var> as if <var>A</var>
+            were assigned the <a>tabindex</a> value 0 for determining its position.</li>
+          <li>Otherwise the <a>sequential focus navigation order</a> for its
+            <a>distributed nodes</a> <strong>must</strong> be inserted in place of
+            <var>A</var> according to its <a>tabindex</a> value.</li>
+        </ol>
+        <p class="note">When a <a>slot</a> <var>A</var> is assigned to another
```
Comment (hayatoito) [inline spec/shadow/index.html]: We do not need to mention this. Please try to have a _general_ rule which covers all cases.

Unless a good general rule, the spec would have a lot of cases which are not covered.
e.g.

``` html
<slot>
  <input></input>
  <input>
    <slot>
       <input><input>
     </slot>
  </input>
</slot>
```

File: spec/shadow/index.html (hunk 2 above)
Comment (hayatoito) [inline spec/shadow/index.html]: Wrong. A's scope is not flattened into B.

File: spec/shadow/index.html [hunk 4]
```diff
@@ -1139,6 +1141,27 @@
...
+        <ol>
+          <li>The <a>sequential focus navigation order</a> for its <a>distributed nodes</a>
+            <strong>must</strong> be skipped if A is assigned <a>tabindex</a> value -1.
+          <li>If <var>A</var> is not given <a>tabindex</a>,
+            The <a>sequential focus navigation order</a> for its <a>distributed nodes</a>
+            <strong>must</strong> be inserted in place of <var>A</var> as if <var>A</var>
+            were assigned the <a>tabindex</a> value 0 for determining its position.</li>
+          <li>Otherwise the <a>sequential focus navigation order</a> for its
+            <a>distributed nodes</a> <strong>must</strong> be inserted in place of
+            <var>A</var> according to its <a>tabindex</a> value.</li>
```
Comment (hayatoito) [inline spec/shadow/index.html]: You might want to clarify what "according to it's tabindex value" means.

A slot element is not focusable element. Thus, you should mention how tabindex value of a slot element is _effectively_ used in focus navigation order more clearly.

File: spec/shadow/index.html (hunk 1 above)
Comment (TakayoshiKochi) [inline spec/shadow/index.html]: Thanks.

Probably we want to start using '[control group](https://html.spec.whatwg.org/multipage/interaction.html#control-group)' (defined in HTML spec) for the scope.
I am working on rewriting with it.

**Comment by TakayoshiKochi** [comment]
@hayatoito PTAL.

I tried to get this upstreamable to HTML spec, [section 6.4](https://html.spec.whatwg.org/multipage/interaction.html#focus),
but it is hard to do in a single step, as the HTML spec describes the
navigation order as an algorithm to iteratively find the next focusable node,
while the Shadow DOM spec (before my modification) assumes there is a
(somewhat) pre-computed _focus navigation order_ for each scope and
describes how they are merged together.

So this is my attempt to include the #375 as exactly as possible by
following the current way of describing focus navigation in Shadow DOM.

At this point it seems to me to patch the upstream spec requires lot more work.

File: spec/shadow/index.html [hunk 5]
```diff
@@ -1123,13 +1123,46 @@
       </section>
 
       <section>
-        <h3>Focus Navigation</h3>
+        <h3>Sequential Focus Navigation</h3>
 
         <p>If a <a>node</a> doesn’t <a data-lt="participates">participate</a> in the <a>document flat tree</a>, the <a>node</a> <strong>must</strong> be skipped from the <a>sequential focus navigation</a>.</p>
 
-        <p>The <a>sequential focus navigation order</a> for a given <a>shadow tree</a> <var>A</var> <strong>must</strong> be inserted into the <a>sequential focus navigation order</a> for the <a>parent tree</a> <var>B</var> as follows:</p>
+        <p>Each <a>shadow root</a> and <a>slot</a> element is a <a>control group owner</a> and
```
Comment (hayatoito) [inline spec/shadow/index.html]: A control group is more upper layer concept, isn't it?

I mean:
- There is 1 : N relationship between [a control group] and [focus navigation order in each scope].
- There is 1 : 1 relationship between [a control group] and [_global_ focus navigation order], which is the result of _merging_ multiple [focus navigation order in each scope].

File: spec/shadow/index.html [hunk 6]
```diff
@@ -1123,13 +1123,46 @@
...
+          The order is tree order.
+        </p>
+
+        <p>For a given <a>slot</a> element,
+          its <a>sequential focus navigation order</a> consists of
+          <a data-lt="focusable area">focusable areas</a> of its <a>assigned nodes</a>
+          and descendants of the <a>assigned nodes</a>.
+          If a given slot's assigned nodes is empty, its
+          <a>sequential focus navigation order</a> consists of
+          <a data-lt="focusable area">focusable areas</a> of its fallback contents.
```
Comment (hayatoito) [inline spec/shadow/index.html]: Fallback contents are not well defined.
Also do not forget that `<slot>` can be a descendant of another `<slot>`. The current definition does not cover this case at all.

**Comment by hayatoito** [comment]
I think you will not get a good definition by the current approach. I recommend you to have a more formal approach:

e.g.

algorithm:
- Input: a document composed tree
- Output: set of owners, where each owner owns a group

algorithm:
- Input: node, NODE, in a document composed tree
- Output: a group which NODE belongs to

Thus, we can define the disjoint sets clearly. Then,

algorithm:
- Input: node A and node B where A nad B joins the same group
- Output: A or B, which comes before in the group

Thus, the _oder_ is clearly defined.

File: spec/shadow/index.html (hunk 5 above)
Comment (TakayoshiKochi) [inline spec/shadow/index.html]: My understanding is that frames own its own control group (e.g. `<iframe>`) and its own sequence,  but from user's standpoint, tab navigation sequence look _global_.  Thus sequences for each group is merged.

**Comment by TakayoshiKochi** [comment]
It looks iterative reviews won't work well if completely different structure is suggested.

I've written the draft following your suggestion:
https://docs.google.com/document/d/1mNV2Z231SrAZuMJvBVNfVNHfULkMwUZOZ75vZM1K9zY/edit?usp=sharing (anyone can comment)

For the 3rd algorithm (ordering) in the comment,
for document/shadow root scope, the order is probably obvious (tree order),
but for slot scope (not fallback content scope) the traversal order is depth-first
search on result of getAssignedNodes()?

In addition, do we want to describe how to compare 2 focusable areas with tabindex here
to get the scoped order of the sequence?

That would add almost dead-copy of [HTML 6.4.3](https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute) and
super redundant.

**Comment by TakayoshiKochi** [comment]
@hayatoito comment on the doc linked above is welcome.

**Comment by TakayoshiKochi** [comment]
@hayatoito
Updated the draft:
https://docs.google.com/document/d/1mNV2Z231SrAZuMJvBVNfVNHfULkMwUZOZ75vZM1K9zY/edit?usp=sharing (anyone can comment)

This should be better aligned with what you suggested... any feedback is welcome.
Sorry for the long turnaround.

**Comment by TakayoshiKochi** [comment]
@hayatoito Updated the draft. PTAL.

**Comment by TakayoshiKochi** [comment]
@hayatoito this branch went into some weird state, so I created another PR #447.
Could you take a look?

**Pull request closed without merging**